File: solutions/java/annalyns-infiltration/2/src/main/java/AnnalynsInfiltration.java
class AnnalynsInfiltration {
    public static boolean canFastAttack(boolean knightIsAwake) {
         return !knightIsAwake;
    }
    
    public static boolean canSpy(boolean knightIsAwake, boolean archerIsAwake, boolean prisonerIsAwake) {
        if(knightIsAwake || archerIsAwake || prisonerIsAwake){
            return true;
        }else{
            return false;
        }
            
    }

    public static boolean canSignalPrisoner(boolean archerIsAwake, boolean prisonerIsAwake) {
        if(prisonerIsAwake && !archerIsAwake){
            return true;
        }else {
            return false;
        }
    }

    public static boolean canFreePrisoner(boolean knightIsAwake, boolean archerIsAwake, boolean prisonerIsAwake, boolean petDogIsPresent) {
        if (petDogIsPresent){
            if (!archerIsAwake){
                return true;
            }else{
                return false;
            }
        }else{
            if (!archerIsAwake && !knightIsAwake && prisonerIsAwake){
                return true;
            }else {
                return false;
            }
            
        }
    }
}

File: solutions/csharp/annalyns-infiltration/1/AnnalynsInfiltration.cs
static class QuestLogic
{
    public static bool CanFastAttack(bool knightIsAwake)
    {
        if(knightIsAwake){
            return false;
        }else{
            return true;
        }
    }

    public static bool CanSpy(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake)
    {
        if (!knightIsAwake && !archerIsAwake && !prisonerIsAwake){
            return false;
        }else{
            return true;
        }
    }

    public static bool CanSignalPrisoner(bool archerIsAwake, bool prisonerIsAwake)
    {
        if(prisonerIsAwake && !archerIsAwake){
            return true;
        }else{
            return false;
        }
    }

    public static bool CanFreePrisoner(bool knightIsAwake, bool archerIsAwake, bool prisonerIsAwake, bool petDogIsPresent)
    {
        if (petDogIsPresent){
            if (!archerIsAwake){
                return true;
            }else{
                return false;
            }
        }else{
            if(prisonerIsAwake && !knightIsAwake && !archerIsAwake){
                return true;
            }else{
                return false;
            }
        }
    }
}
